### PR TITLE
Properly setting and unsetting the disabled state for juniper switches

### PR DESCRIPTION
**Pull request by idjaw**

### CODE REVIEW

File: netman/adapters/switches/juniper/base.py [hunk 1]
```diff
@@ -425,7 +425,15 @@ def set_interface_state(self, interface_id, state):
             self.logger.info("actual setting error was {}".format(e))
             raise UnknownInterface(interface_id)
 
+    def reset_interface_state(self, interface_id):
+        update = Update()
+        update.add_interface(interface_main_update(interface_id, None))
```
Comment (mat128) [inline netman/adapters/switches/juniper/base.py]: I don't see how this changes the state, is the second parameter to `interface_main_update` the state? I suggest renaming the method and/or providing kwargs to make this clearer.

Comment (idjaw) [inline netman/adapters/switches/juniper/base.py]: Agreed. Will refactor.

File: tests/adapters/switches/juniper_test.py [hunk 2]
```diff
@@ -4297,6 +4297,68 @@ def test_enable_interface_succeeds(self):
...
+                <interfaces>
+                  <interface>
+                    <name>ge-0/0/6</name>
+                  </interface>
+                </interfaces>
+              </configuration>
+            </config>
+        """)).and_return(an_ok_response())
+
+        self.switch.reset_interface_state("ge-0/0/6")
```
Comment (lindycoder) [inline tests/adapters/switches/juniper_test.py]: this is NOT how juniper works, at all

first of all : there's no need to do these 2 commands in a unit test if they are mocked.  there's no state to the mock

second : to remove a member from the tree you need to give it operation="delete"

See unset_interface_native_vlan for reference

File: tests/adapters/switches/juniper_test.py [hunk 3]
```diff
@@ -4297,6 +4297,68 @@ def test_enable_interface_succeeds(self):
...
 
+    def test_reset_interface_succeeds_removing_enabled_state(self):
+
+        self.netconf_mock.should_receive("edit_config").once().with_args(target="candidate", config=is_xml("""
+            <config>
+              <configuration>
+                <interfaces>
+                  <interface>
+                    <name>ge-0/0/6</name>
+                    <enable />
```
Comment (lindycoder) [inline tests/adapters/switches/juniper_test.py]: it has come to our ears that the enable here is not a real attribute is it is faulty to use it, therefore this method should only remove a <disable> node and i think you need to be sure it is there, please verify

File: netman/adapters/switches/juniper/base.py [hunk 4]
```diff
@@ -425,7 +425,15 @@ def set_interface_state(self, interface_id, state):
             self.logger.info("actual setting error was {}".format(e))
             raise UnknownInterface(interface_id)
 
+    def reset_interface_state(self, interface_id):
```
Comment (lindycoder) [inline netman/adapters/switches/juniper/base.py]: Naming convention is unset_interface_state

File: netman/adapters/switches/juniper/base.py [hunk 5]
```diff
@@ -413,11 +413,9 @@ def edit_interface_spanning_tree(self, interface_id, edge=None):
 
                 self._push(update)
 
-    def set_interface_state(self, interface_id, state):
+    def set_interface_state(self, interface_id, state=ON):
```
Comment (lindycoder) [inline netman/adapters/switches/juniper/base.py]: This does not respect the parent method's signature

File: netman/adapters/switches/juniper/base.py [hunk 6]
```diff
@@ -425,7 +423,8 @@ def set_interface_state(self, interface_id, state):
             self.logger.info("actual setting error was {}".format(e))
             raise UnknownInterface(interface_id)
 
-
+    def unset_interface_state(self, interface_id):
+        self.set_interface_state(interface_id)
```
Comment (lindycoder) [inline netman/adapters/switches/juniper/base.py]: please provide the state=ON here instead

File: tests/adapters/switches/juniper_test.py [hunk 7]
```diff
@@ -4102,48 +4102,65 @@ def test_enable_interface_succeeds(self):
...
+            <error-severity>error</error-severity>
+            <error-message>
+            port value outside range 0..47 for '99' in 'ge-0/0/99'
+            </error-message>
+            </rpc-error>"""))))
+
+        with self.assertRaises(UnknownInterface) as expect:
+            self.switch.set_interface_state("ge-0/0/99", ON)
+
+        assert_that(str(expect.exception), contains_string("Unknown interface ge-0/0/99"))
```
Comment (lindycoder) [inline tests/adapters/switches/juniper_test.py]: missing test: test_unset_interface_state_unknown_interface_raises

**Comment by idjaw** [comment]
Fixed code per comments.

File: tests/adapters/switches/juniper_test.py [hunk 8]
```diff
@@ -5352,6 +5394,7 @@ def compare(self, expected, actual):
                 else:
                     assert_that(actual[i].text is not None, "Node is " + node.tag)
                     assert_that(node.text.strip(), equal_to(actual[i].text.strip()))
+            assert_that(actual[i].attrib, has_length(len(node.attrib)))
```
Comment (fbouliane) [inline tests/adapters/switches/juniper_test.py]: Maybe Move to [5390](https://github.com/internap/netman/pull/107/files#diff-034dde710188c9228ebdc20f875cf05bR5390) with his friends !

File: netman/adapters/switches/juniper/base.py [hunk 9]
```diff
@@ -415,9 +415,7 @@ def edit_interface_spanning_tree(self, interface_id, edge=None):
 
     def set_interface_state(self, interface_id, state):
         update = Update()
-        update.add_interface(interface_main_update(interface_id, [
-            to_ele("<disable />") if state is OFF else to_ele("<enable />")
-        ]))
+        update.add_interface(interface_state_update(interface_id, state))
```
Comment (fbouliane) [inline netman/adapters/switches/juniper/base.py]: Before it did :

    <interface>
        <name>Ge-0/0/99</name>
    </interface>
    <disable />

And now it does :

    <interface>
        <name>Ge-0/0/99</name>
        <disable />
    </interface>

?

Comment (idjaw) [inline netman/adapters/switches/juniper/base.py]: If you are setting the state to off then yes, `<disable />` will be in place. If not, then an operation to attempt to remove `<disable />` will be set per: `<disable operation="delete" />`

File: tests/adapters/switches/juniper_test.py (hunk 8 above)
Comment (idjaw) [inline tests/adapters/switches/juniper_test.py]: wow.....yeah.

Comment (idjaw) [inline tests/adapters/switches/juniper_test.py]: tests still passed too. Winning.

File: netman/adapters/switches/juniper/base.py (hunk 9 above)
Comment (fbouliane) [inline netman/adapters/switches/juniper/base.py]: Ok, .append not append at then  end but inside the element, so behavior has not changed, Thanks.

**Pull request closed without merging**